Optionally adding extra information to meta.json.

terjekv opened the pull request:
This patch seeks to add some useful system information to the meta.json file for logging and monitoring. This feature is optional.

  - Adds the following fields to meta.json:
    - cvmfs_version : version_string
    - os_version : $(lsb_release -d | cut -f2)
  - Can be disabled by setting PUBLISH_VERSIONS_IN_META_FILE to anything but "true".

terjekv committed:
Optionally adding extra information to meta.json.

This patch seeks to add some useful system information to the meta.json file for logging and monitoring. This feature is optional.

  - Adds the following fields to meta.json:
    - cvmfs_version : version_string
    - os_version : $(lsb_release -d | cut -f2)
  - Can be disabled by setting PUBLISH_VERSIONS_IN_META_FILE to anything but "true".

terjekv committed:
Remove rogue whitespace.

terjekv commented:
As stated on Mattermost, I find it quite tricky to test this in a nice fashion. Suggestions on how to test this locally would be greatly appreciated. :)

radupopescu commented:
@cernvm-bot tidy

radupopescu commented:
@cernvm-bot cpplint

radupopescu commented:
@cernvm-bot unittest

radupopescu commented:
@cernvm-bot cloudtest

radupopescu commented on cvmfs/server/cvmfs_server_coda.sh:
```diff
@@ -23,6 +23,10 @@ DEFAULT_LOCAL_STORAGE="/srv/cvmfs"
 
+# Should we publish CVMFS and OS versions in meta.json?
+# Set to any non-true string to turn off this feature.
+PUBLISH_VERSIONS_IN_META_FILE=true
```
As this variable is supposed to be overridable in the repository configuration files, we should be consistent and rename it to start with `CVMFS_`.

terjekv replied:
Yup, good point!

radupopescu commented on cvmfs/server/cvmfs_server_coda.sh:
```diff
@@ -23,6 +23,10 @@ DEFAULT_LOCAL_STORAGE="/srv/cvmfs"
 
+# Should we publish CVMFS and OS versions in meta.json?
+# Set to any non-true string to turn off this feature.
+PUBLISH_VERSIONS_IN_META_FILE=true
```
```suggestion
CVMFS_PUBLISH_VERSIONS_IN_META_FILE=true
```

radupopescu commented on cvmfs/server/cvmfs_server_json.sh:
```diff
@@ -91,6 +91,11 @@ _render_info_file() {
     echo '  "last_geodb_update" : "'$modtime'",'
   fi
 
+  if [ "x${PUBLISH_VERSIONS_IN_META_FILE}" == "xtrue" ]; then
```
```suggestion
  if [ "x${CVMFS_PUBLISH_VERSIONS_IN_META_FILE}" == "xtrue" ]; then
```

terjekv committed:
Update cvmfs/server/cvmfs_server_coda.sh

Co-authored-by: Radu Popescu <mail@radupopescu.net>

terjekv committed:
Update cvmfs/server/cvmfs_server_json.sh

Co-authored-by: Radu Popescu <mail@radupopescu.net>

terjekv commented:
Huh, https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsCloudTesting/1354/CVMFS_PLATFORM=el7,CVMFS_PLATFORM_CONFIG=x86_64,label=trampoline/testReport/(root)/ServerIntegrationTests/cvmfs_server_check__a/ is weird. 

This doesn't help me much. :( 

`````
Starting test.cern.ch.d5a293.replica-4 at Tue Mar 15 15:32:05 CET 2022
Verifying integrity of test.cern.ch.d5a293.replica-4...
failed to fetch repository manifest (8 - bad whitelist)
failed to load repository manifest
ERROR from cvmfs_server check!
Finished test.cern.ch.d5a293.replica-4 at Tue Mar 15 15:32:05 CET 2022
*** checking if check worked on test repo and replicas 1 and 3
  "check_status": "succeeded"
  "check_status": "succeeded"
  "check_status": "succeeded"
*** verifying check did not run on replica 2
*** verifying check failed on replica 4
  "check_status": "failed"
*** verify check was delayed on replica 1
`````

jblomer commented on cvmfs/server/cvmfs_server_json.sh:
```diff
@@ -91,6 +91,11 @@ _render_info_file() {
 
+  if [ "x${CVMFS_PUBLISH_VERSIONS_IN_META_FILE}" == "xtrue" ]; then
+    echo '  "cvmfs_version" : "$(cvmfs_version_string)",'
+    echo '  "os_version" : "' $(lsb_release -d | cut -f2) '", '
```
For this we'd need to add the `lsb-release` dependency to the cvmfs-server package.

DrDaveD replied:
And watch out, "there are no plans to ship lsb-release in RHEL9".  Should use /etc/os-release instead.

terjekv replied:
At some point I will understand what the community decides on as a dependable, long term, consistent way of figuring out the running version of linux... I'm not sure we can depend on `/etc/os-release` as it is theoretically tied to systemd, and as much as Debians and RHEL clones have it, god knows for how long? I am happy to use it though if the CVMFS community prefers it they way, but it should probably be moved to a function somewhere ala `_get_os_description` or similar.

Other options is to drop the data, or doing `uname -r`, but `uname` leaks data we probably don't want to announce and doesn't really give me what I want anyway. Hm. I'm very open to ideas here.

terjekv replied:
With regards to `/etc/os-release`, the default way to parse the file is to source it, ala:
````
test -e /etc/os-release && os_release='/etc/os-release' || os_release='/usr/lib/os-release'
. "${os_release}"
````
However, this pollutes the environment of the running script with environment variables like `NAME`, `ID`, `HOME_URL`, and more. The environment variables used in `cvmfs_server` are all sanely prefixed and should avoid being clobbered, but I'm still a bit nervous about doing a raw source late in the scripts run time. 

My thought would be to either parse the file, or if sourcing it, doing it very early in https://github.com/cvmfs/cvmfs/blob/devel/cvmfs/server/cvmfs_server_coda.sh to ensure any clobbering is done by CMVFS code and not from `/etc/os-release`.

terjekv replied:
Implemented one approach in https://github.com/cvmfs/cvmfs/pull/2863/commits/03bbc163cf6e41cb8ecec8574dd4fb0263f244fd

jblomer reviewed (changes requested):
Many thanks for the contribution!

terjekv committed:
Use {/etc,/var/lib}/os-release for os information.

  - Be permissive!
  - If os-release isn't found, return empty strings.
  - If fields requested aren't found, return emtpy strings.

terjekv committed:
Fix interpolation of cvmfs_version_string.

amadio commented on cvmfs/server/cvmfs_server_util.sh:
```diff
@@ -733,6 +733,54 @@ has_selinux() {
+		# If we are unable to find a proper /etc/os-release file return nothing.
+		:
+	else
+		grep "^${fieldname}=" ${_CVMFS_OS_RELEASE_FILENAME} | cut -f2 -d'"'
```
Suggestion: `eval $(sed -n "/^${fieldname}=/p" ${_CVMFS_OS_RELEASE_FILENAME})`

terjekv replied:
Huh. Interesting approach. Saves a pipe but introduces an eval. Is there a specific reason for wanting this approach? I don't mind either way, I'm mostly asking to learn something. :)

amadio replied:
Actually, nevermind, the eval will add the variable to the environment, and you said this is something you don't want, right? Otherwise you could just source the os-release file... I think the way you did is better, but a single sed can probably replace the grep + cut indeed.

DrDaveD replied:
The grep..cut isn't going to work if the double-quotes are missing.  I see that is the case on debian for ID, where it is `ID=debian`.

On the other hand, the eval does not print to stdout as expected.  This one works for me:
```
eval echo $(sed -n "s/^${fieldname}=//p" /etc/os-release)
```

amadio replied:
You don't need the eval in that case, just do `sed -n "s/^${fieldname}=//p" /etc/os-release` (you may get an extra newline like that, though).

DrDaveD replied:
I tried that, but it doesn't remove the double quotes ... but you could do that with
```
sed -n "s/\"//g;s/^${fieldname}=//p" /etc/os-release
```

DrDaveD replied:
Oh I thought that looked familiar.  I wrote that same line [before](https://github.com/cvmfs-contrib/config-repo/blob/osg/libexec/osgstorage-auth.conf#L20).

terjekv replied:
Aha. :) https://github.com/cvmfs/cvmfs/pull/2863/commits/703dad1b9603d5fca9ee3d6359e3f5496f79c0ca now uses the line from @DrDaveD. Do note that supposedly the os-release-file may be located under `/usr/lib`.

DrDaveD commented on cvmfs/server/cvmfs_server_util.sh:
```diff
@@ -733,6 +733,54 @@ has_selinux() {
+_os_etc_release_get_field() {
+	_os_set_etc_release_filename
+
+	fieldname=$1
```
```suggestion
	local fieldname=$1
```

terjekv committed:
Update cvmfs/server/cvmfs_server_util.sh

Co-authored-by: DrDaveD <2129743+DrDaveD@users.noreply.github.com>

terjekv committed:
Support non-quoted field values in the OS release file.

jblomer commented on cvmfs/server/cvmfs_server_util.sh:
```diff
@@ -733,6 +733,54 @@ has_selinux() {
+		# If we are unable to find a proper /etc/os-release file return nothing.
+		:
+	else
+    sed -n "s/\"//g;s/^${fieldname}=//p" ${_CVMFS_OS_RELEASE_FILENAME}
```
There seems to be a whitespacing problem here. Perhaps a mix of tabs and spaces? It should be spaces only.

terjekv replied:
Whops. Sorry. Fixed in https://github.com/cvmfs/cvmfs/pull/2863/commits/e5150a49ba45a423259a3397740734f0bf8bef03

jblomer reviewed (changes requested):
One tiny issue, otherwise it looks fine to me.  Many thanks for the PR!

terjekv committed:
Ensure whitespace is spaces only.

jblomer commented:
@cernvm-bot cloudtest

terjekv commented:
> cloudtests finished: FAILURE
> https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsCloudTesting/1358/

This again does stuff like this:

````
Starting test.cern.ch.21518a.replica-4 at Thu Mar 17 10:08:20 CET 2022
Verifying integrity of test.cern.ch.21518a.replica-4...
failed to fetch repository manifest (8 - bad whitelist)
failed to load repository manifest
ERROR from cvmfs_server check!
Finished test.cern.ch.21518a.replica-4 at Thu Mar 17 10:08:20 CET 2022
*** checking if check worked on test repo and replicas 1 and 3
  "check_status": "succeeded"
  "check_status": "succeeded"
  "check_status": "succeeded"
*** verifying check did not run on replica 2
*** verifying check failed on replica 4
  "check_status": "failed"
*** verify check was delayed on replica 1
````

Is this expected or something I've messed up?

jblomer commented:
> Is this expected or something I've messed up?

That may be a problem due to the patch. Possibly just an issue with the test itself. We'd need to look into it, it may take a few days until I get to it.

terjekv commented:
Ah, thanks! I'll wait to see how things go. :)

terjekv commented:
@jblomer Did anyone find out what caused the tests to fail for this PR?

jblomer commented:
Apologies for the delay! We need to fix the test, the problem shows with and without your patch.

I think it's good to go, we'll merge it tomorrow.

terjekv commented:
No worries, thank you for following up on this issue!

terjekv commented:
I noticed this didn't make it into 2.9.4. Are we looking at 2.10 for extra features such as this?

jblomer commented:
That's right, by default we merge into the devel branch, which becomes the next feature release (in this case: 2.10). Fixes for the patch releases need to be explicitly backported.  For 2.10, we plan to cut a release for the CernVM workshop in September.  We may be looking at a prelease, so ETA for 2.10 is 1-2 months.

terjekv commented:
Cool, thanks!
